### ISSUE TITLE
 way/:id/full/history

### ISSUE THREAD

**Issue by emacsen** [openstreetmap/openstreetmap-website#130]
Trying to look at the history of the geometry of a way is difficult. Doing so means that a client would need to call a way's history, then retrieve node history for each node.

This means a lot of requests to the server.

If the server could return a full history that included all the versions of all the way, along with all the history of the nodes ever in that way, this would make generating a geometry history much easier and less expensive to the server in terms of reducing the number of calls.

**Comment by mmd-osm** [comment]
Forthcoming Overpass API 0.7.55 has a new feature to return a way's full history (at least back to Sep 2012), see: https://github.com/drolbr/Overpass-API/issues/282#issuecomment-315516184

Example: http://overpass-turbo.eu/s/uTE

**Comment by tyrasd** [comment]
> Forthcoming Overpass API 0.7.55 has a new feature to return a way's full history

as far as I can see, this still doesn't perfectly fulfill what's been requested in this thread:

> […] return a full history that included all the versions of all the way, along with all the history of the nodes ever in that way […]

While (AFAICS) this new Overpass-API functionality "only" returns _all the versions of the way, along with resolved nodes of this way at the way's versions' timestamps_.

**Comment by mmd-osm** [comment]
@tyrasd : yes, that's correct. If you just move around some nodes, the way won't get a new version, and consequently that change is not reflected in the query result.

I will check with Roland if `timeline` can also be used more generically by providing a list of nodes in an inputset, rather than via an explicit object id as in my example. Then you could use a nested `timeline` / `foreach` / `retro` query.

**Comment by mmd-osm** [comment]
@Zverik has brought up this topic again in http://shtosm.ru/all/zhelaem-togo-chto-imeem/ and https://2019.stateofthemap.org/sessions/V7NUWP/

This could be done by a single SQL statement (see https://gist.github.com/mmd-osm/fff7f5620def9affd5477da6aeb419b9 for how this could look like for ways). I don't know how expensive this is in production, and if we should restrict it to a range of version numbers rather than evaluating the full history.

The example query returns a flat list of node id/version numbers, which would be served back to the client along with the respective way versions (which is a separate SQL statement). This way, the amount of data to be transferred would be kept to an absolute minimum. It would be the responsibility of the client to reassemble the full history based on those node and way versions, which shouldn't really be an issue.

**Comment by mmd-osm** [comment]
Adding to https://github.com/openstreetmap/openstreetmap-website/pull/1448: if we restrict `way/:id/full/history` to the previous and current object version, we would automatically get all the relevant details to render a representation of a deleted object at the time it was deleted.

**Comment by mmd-osm** [comment]
I put this up on the dev instance for some very early testing. The endpoint URL is different from what has been proposed here, so we don't have to change any of the Apache rewrite rules.

Some examples:

- https://upload.apis.dev.openstreetmap.org/api/0.6/way/180401/history?nodes
- https://upload.apis.dev.openstreetmap.org/api/0.6/way/272908/history?nodes

If you like you can create your own account on https://upload.apis.dev.openstreetmap.org and try for yourself.
